File: Backend/soundwave/src/main/java/org/example/soundwave/service/TypeService.java
package org.example.soundwave.service;

import lombok.RequiredArgsConstructor;
import org.example.soundwave.model.dto.BrandDTO;
import org.example.soundwave.model.dto.TypeDTO;
import org.example.soundwave.model.entity.Brand;
import org.example.soundwave.model.entity.Type;
import org.example.soundwave.model.exception.TypeException;
import org.example.soundwave.model.request.TypeRequest;
import org.example.soundwave.model.response.PageResponse;
import org.example.soundwave.repository.TypeRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class TypeService {
    private final TypeRepository typeRepository;

    public Type findTypeById(Long id){
        return typeRepository.findTypeById(id)
                .orElseThrow(() -> new TypeException("Type with id: " + id + " do not exist"));
    }

    public Type findTypeByName(String typeName) {
        return typeRepository.findTypeByName(typeName)
                .orElseThrow(() -> new TypeException("Type with name: " + typeName + " do not exist"));
    }

    public PageResponse<TypeDTO> getTypes(int pageNo, int pageSize, String sortBy, String sortDir, String name) {
        Sort sort = sortDir.equalsIgnoreCase(Sort.Direction.ASC.name()) ?
                Sort.by(sortBy).ascending() : Sort.by(sortBy).descending();

        Pageable pageable = PageRequest.of(pageNo, pageSize, sort);
        Page<Type> types = typeRepository.findByNameContainingIgnoreCase(name, pageable);

        List<TypeDTO> content = types.getContent()
                .stream()
                .map(TypeDTO::new)
                .collect(Collectors.toList());

        return new PageResponse<>(
                content,
                types.getNumber(),
                types.getSize(),
                types.getTotalElements(),
                types.getTotalPages(),
                types.isLast());
    }

    public List<TypeDTO> getAllTypes() {
        return typeRepository.findAll()
                .stream()
                .map(TypeDTO::new)
                .collect(Collectors.toList());
    }

    public void addType(TypeRequest request) {
        Type type = Type.builder()
                .name(request.name()).build();

        typeRepository.save(type);
    }

    public void editType(Long id, TypeRequest request) {
        Type type = findTypeById(id);

        type.setName(request.name());

        typeRepository.save(type);
    }

    public void deleteType(Long id) {
        Type type = findTypeById(id);

        typeRepository.delete(type);
    }
}

File: Frontend/soundwave/src/services/typeService.js
import {apiClient, handleApiError} from './apiService.js';

export const typeService = {
    async getTypes(params = {}) {
        const {
            page = 0,
            size = 10,
            sortBy = 'id',
            sortDir = 'asc',
            name = ''
        } = params;

        const queryParams = new URLSearchParams();
        queryParams.append('page', page.toString());
        queryParams.append('size', size.toString());
        queryParams.append('sortBy', sortBy);
        queryParams.append('sortDir', sortDir);
        queryParams.append('name', name);

        try {
            const response = await apiClient.get(`/api/type?${queryParams.toString()}`);
            return response.data;
        } catch (error) {
            handleApiError(error, 'Failed to fetch types');
        }
    },

    async getAllTypes() {
        try {
            const response = await apiClient.get('/api/type/all');
            return response.data;
        } catch (error) {
            handleApiError(error, 'Failed to fetch types');
        }
    },

    async addType(typeRequest) {
        try {
            const response = await apiClient.post('/api/type/add', typeRequest);
            return response.data;
        } catch (error) {
            handleApiError(error, 'Failed to add type');
        }
    },

    async editType(id, typeRequest) {
        try {
            const response = await apiClient.put(`/api/type/${id}`, typeRequest);
            return response.data;
        } catch (error) {
            handleApiError(error, 'Failed to edit type');
        }
    },

    async deleteType(id) {
        try {
            const response = await apiClient.delete(`/api/type/${id}`);
            return response.data;
        } catch (error) {
            handleApiError(error, 'Failed to delete type');
        }
    }
};

export default typeService;